File: problem-solving-rb/is-it-a-binary-search-tree.rb
class Node
	attr_reader :data
	attr_accessor :left, :right
	
	def initialize(data)
		@data = data
	end
end

def array_to_tree(array, i = 0)
  # use your function from the previous challenge
  return nil if i >= array.length || array[i].zero?

  node = Node.new(array[i])
  node.left = array_to_tree(array, 2 * i + 1)
  node.right = array_to_tree(array, 2 * i + 2)

  node
end

def check_bst?(temp, min = nil, max = nil)
  return true if temp.nil?

  if (min && temp.data <= min) || (max && temp.data >= max)
    return false
  end

  left = check_bst?(temp.left, min, temp.data)
  right = check_bst?(temp.right, temp.data, max)

  left && right
end

def search_tree?(array)
  root = array_to_tree(array)
  # write your code here
  check_bst?(root)
end

puts search_tree?([10, 4, 12])
# => true

puts search_tree?([10, 5, 7])
# => false

File: problem-solving-js/is-it-a-binary-search-tree.js
class Node {
  constructor(data) {
    this.data = data;
    this.left = null;
    this.right = null;
  }
}

function arrayToTree(array, index = 0) {
  if (index >= array.length || array[index] === null) {
    return null;
  }

  const node = new Node(array[index]);
  node.left = arrayToTree(array, 2 * index + 1);
  node.right = arrayToTree(array, 2 * index + 2);

  return node;
}

function validBST(node, min = null, max = null) {
  if (node === null) {
    return true;
  }

  if ((min && node.data <= min) || (max && node.data >= max)) {
    return false;
  }

  return (
    validBST(node.left, min, node.data) && validBST(node.right, node.data, max)
  );
}

function searchTree(array) {
  const root = arrayToTree(array);
  return validBST(root, null, null);
}

console.log(searchTree([10, 4, 12])); // Output: true
console.log(searchTree([10, 5, 7])); // Output: false
